### PR TITLE
swarm/network: fix data race in stream.(*Peer).handleOfferedHashesMsg()

### PR DESCRIPTION
**Pull request by frncmx**
handleOfferedHashesMsg() contained a data race:
- read => in a goroutine, call to c.batchDone()
- write => in the main thread, write to c.sessionAt

c.batchDone() contained a call to c.AddInterval(). Client was a value
receiver for AddInterval. So on c.AddInterval() call the whole client
struct got copied (read) while one of its field was modified in
handleOfferedHashesMsg() (write).

fixes ethersphere#1086

### CODE REVIEW

File: swarm/network/stream/stream.go [hunk 1]
```diff
@@ -666,17 +666,16 @@ func peerStreamIntervalsKey(p *Peer, s Stream) string {
...
+	if err = c.intervalsStore.Get(c.intervalsKey, i); err != nil {
 		return err
 	}
 	i.Add(start, end)
 	return c.intervalsStore.Put(c.intervalsKey, i)
 }
 
-func (c client) NextInterval() (start, end uint64, err error) {
+func (c *client) NextInterval() (start, end uint64, err error) {
 	i := &intervals.Intervals{}
```
Comment (janos) [inline swarm/network/stream/stream.go]: As intervals.Intervals allocation is changed in AddIntervals to use new, and the Get function call is inside if, should this be changed here, also?

Comment (frncmx) [reply swarm/network/stream/stream.go]: Nice! :1st_place_medal:  
Yup. Let's keep it consistent.  :)

Comment (frncmx) [reply swarm/network/stream/stream.go]: fixed

Comment (janos) [reply swarm/network/stream/stream.go]: Cool, even though, I do not see why this change is needed to solve the issue, but if you like to make it better, great. There are other examples, like https://github.com/ethereum/go-ethereum/blob/master/swarm/network/stream/stream.go#L112, https://github.com/ethereum/go-ethereum/blob/master/swarm/network/stream/peer.go#L329.

File: swarm/network/stream/stream.go [hunk 2]
```diff
@@ -666,18 +666,17 @@ func peerStreamIntervalsKey(p *Peer, s Stream) string {
...
 		return err
 	}
 	i.Add(start, end)
 	return c.intervalsStore.Put(c.intervalsKey, i)
 }
 
-func (c client) NextInterval() (start, end uint64, err error) {
-	i := &intervals.Intervals{}
+func (c *client) NextInterval() (start, end uint64, err error) {
+	i := new(intervals.Intervals)
```
Comment (nonsense) [inline swarm/network/stream/stream.go]: Nitpick, but this is not the style that most Go programs use. Generally you create a new pointer value with `&smth{}` and not `new(smth)`. I am pretty sure this is much more common across our codebase as well.

File: swarm/network/stream/stream.go [hunk 3]
```diff
@@ -666,18 +666,17 @@ func peerStreamIntervalsKey(p *Peer, s Stream) string {
 	return p.ID().String() + s.String()
 }
 
-func (c client) AddInterval(start, end uint64) (err error) {
-	i := &intervals.Intervals{}
-	err = c.intervalsStore.Get(c.intervalsKey, i)
-	if err != nil {
+func (c *client) AddInterval(start, end uint64) (err error) {
+	i := new(intervals.Intervals)
```
Comment (nonsense) [inline swarm/network/stream/stream.go]: Same comment as above. I don't think we should be changing the style.

File: swarm/network/stream/stream.go (hunk 2 above)
Comment (frncmx) [reply swarm/network/stream/stream.go]: solved

File: swarm/network/stream/stream.go (hunk 1 above)
Comment (janos) [reply swarm/network/stream/stream.go]: Bot allocations a re fine for me, it is good to be consistent as much as possible. Shouldn't the err assignment a line bellow be under the if statement just as it is changed in AddInterval?

Comment (frncmx) [reply swarm/network/stream/stream.go]: I strongly prefer consistency, too. My eyes skipped that other `err` statement. Too late I guess. :(

**Pull request merged**